### PR TITLE
Feature/change remote address

### PR DESCRIPTION
**Pull request by jthomas43**

### CODE REVIEW

File: include/udx.h [hunk 1]
```diff
@@ -109,6 +109,7 @@ typedef void (*udx_socket_close_cb)(udx_socket_t *handle);
 typedef int (*udx_stream_firewall_cb)(udx_stream_t *handle, udx_socket_t *socket, const struct sockaddr *from);
 typedef void (*udx_stream_read_cb)(udx_stream_t *handle, ssize_t read_len, const uv_buf_t *buf);
 typedef void (*udx_stream_drain_cb)(udx_stream_t *handle);
+typedef void (*udx_stream_remote_changed_cb)(udx_stream_t *stream);
```
Comment (kasperisager) [inline include/udx.h:112]:
```suggestion
typedef void (*udx_stream_remote_changed_cb)(udx_stream_t *handle);
```

File: include/udx.h [hunk 2]
```diff
@@ -399,6 +404,9 @@ udx_stream_set_ack (udx_stream_t *handle, uint32_t ack);
 int
 udx_stream_connect (udx_stream_t *handle, udx_socket_t *socket, uint32_t remote_id, const struct sockaddr *remote_addr);
 
+int
+udx_stream_change_remote (udx_stream_t *stream, uint32_t remote_id, const struct sockaddr *remote_addr, udx_stream_remote_changed_cb remote_change_cb);
```
Comment (kasperisager) [inline include/udx.h:408]:
```suggestion
udx_stream_change_remote (udx_stream_t *stream, uint32_t remote_id, const struct sockaddr *remote_addr, udx_stream_remote_changed_cb remote_changed_cb);
```

File: src/udx.c [hunk 3]
```diff
@@ -909,6 +909,7 @@ rack_detect_loss (udx_stream_t *stream) {
 
 static void
 ack_update (udx_stream_t *stream, uint32_t acked, bool is_limited) {
+
```
Comment (kasperisager) [inline src/udx.c:912]:
```suggestion
```

File: src/udx.c [hunk 4]
```diff
@@ -1888,6 +1901,59 @@ udx_stream_check_timeouts (udx_stream_t *handle) {
...
+    return UV_EINVAL;
+  }
+
+  memcpy(&stream->remote_addr, remote_addr, stream->remote_addr_len);
+  stream->remote_id = remote_id;
+
+  printf("stream->seq=%d stream->remote_acked=%d\n", stream->seq, stream->remote_acked);
+
+  if (stream->seq != stream->remote_acked) {
+
```
Comment (kasperisager) [inline src/udx.c:1931]:
```suggestion
```

File: src/udx.c [hunk 5]
```diff
@@ -1888,6 +1901,59 @@ udx_stream_check_timeouts (udx_stream_t *handle) {
...
+    stream->remote_changing = true;
+    stream->seq_on_remote_changed = stream->seq;
+    stream->on_remote_changed = on_remote_changed;
+  } else {
+
+    printf("ready to change now\n");
```
Comment (kasperisager) [inline src/udx.c:1937]:
```suggestion
```

File: src/udx.c [hunk 6]
```diff
@@ -1888,6 +1901,59 @@ udx_stream_check_timeouts (udx_stream_t *handle) {
...
+  stream->mtu_probe_count = 0;
+  stream->mtu_probe_size = UDX_MTU_BASE; // starts with first ack, counts as a confirmation of base
+  stream->mtu_max = UDX_MTU_MAX;         // revised in connect()
+
+  uv_timer_stop(&stream->mtu_raise_timer);
+
+  if (stream->socket->family == 6 && stream->remote_addr.ss_family == AF_INET) {
+    addr_to_v6((struct sockaddr_in *) &stream->remote_addr);
+    stream->remote_addr_len = sizeof(struct sockaddr_in6);
+  }
```
Comment (kasperisager) [inline src/udx.c:1952]: Should this be moved to _before_ the `on_remote_changed` callback is invoked just to be safe?

File: test/stream-change-remote.c [hunk 7]
```diff
@@ -0,0 +1,126 @@
...
+ * r = socket(AF_INET, SOCK_DGRAM)
+ * r.bind(("127.0.0.2", 8082))
+ *
+ * s = socket(AF_INET, SOCK_DGRAM)
+ * s.connect(("127.0.0.1", 8082))
+ *
+ * while True:
+ *     buf = r.recv(2000)
+ *     s.send(buf)
+ */
```
Comment (kasperisager) [inline test/stream-change-remote.c:15]: You can use the `udx_stream_relay_to()` function for setting up similar relaying with just UDX streams.

File: src/udx.c (hunk 6 above)
Comment (jthomas43) [reply src/udx.c]: suggestion taken

File: test/stream-change-remote.c (hunk 7 above)
Comment (jthomas43) [reply test/stream-change-remote.c]: suggestion taken, test updated

**Review by kasperisager — APPROVED** [review]
Looks solid! 👏

**Pull request merged**